WordPress/gutenberg
# Experiment: Auto-inserting blocks on the frontend and in the editor (via REST API)

## What?
Based on concepts discussed in https://github.com/WordPress/gutenberg/issues/39439, this PR explores auto-inserting blocks in the editor by means of the REST API.

## Why?
We've discussed multiple ways of auto-inserting blocks in the editor, and the REST API route seemed promising enough to try it out. For more details, see #39439.

## How?
Unlike the frontend where auto-insertion can happen at the render stage (e.g. via the `render_block` and `render_block_data` filters, see #50103 and #51294), things are quite different for the REST API:

- It's much less monolithic. To edit a template in the editor, the client will fetch templates (and template parts) from the `/templates` endpoint, block patterns from the `/patterns` endpoint, etc. This means that we need to identify all such possible sources, and make sure to implement auto-insertion for all of them.
- There's no rendering stage, at least not in the templates endpoint; instead, we're serving serialized block markup. This means that if we want to auto-insert blocks, we need to parse that markup, insert our blocks, and re-serialize it. Fortunately, there are a few filters that we can leverage here, e.g. `get_block_file_template` and `get_block_templates`.
  - Of note, we're already (conditionally) doing the parsing/modifying/reserializing dance in `_inject_theme_attribute_in_block_template_content` and `_remove_theme_attribute_in_block_template_content`. In the long run, we should try to provide some specialized hook to give filters access to the parsed blocks before reserialization to avoid the performance impact from doing that dance more than once.
- It's possible to use the same auto-insertion logic (i.e. right after block parsing) for the frontend. This is probably beneficial as it avoids duplicating the relevant code.

Note that this PR is branched from #50103.

## Testing Instructions

See the video below for a walkthrough.

- Enable the "Auto-inserting blocks" experiment on the Experiments page.
- Install the [Like Button example plugin](https://github.com/ockham/like-button).
  - You can either download a plugin zip from the Releases tab, or clone the git repo and link it from your Gutenberg dev env, e.g. via a `.wp-env.override.json` file.
 - Alternatively, you can also add an `__experimentalAutoInsert` field to the `block.json` of a dynamic block of your liking. You have to specify the "anchor" block and the relative position for auto-insertion:

```json
	"__experimentalAutoInsert": {
		"core/comment-template": "lastChild"
	}
```

(A `gutenberg_register_auto_inserted_block()` function is also provided as an alternative.)

- View any page that contains the anchor block, and verify that the auto-inserted block is, indeed, automatically inserted 🙂 _(Make sure that the template that's responsible for rendering that page doesn't have any user customizations though!)_
- Open the corresponding template in the Site Editor. Verify that the block is also auto-inserted here.
- Edit the template in whatever way (e.g. moving or removing the auto-inserted block or the anchor block), and verify that the frontend reflects your edits.

## Screenshots or screencast

https://www.loom.com/share/7b2cb29672ce4173b3844ad836de09d9?sid=4357767c-9a67-470a-8820-a1eff0fbfab7


## Q&A

**What happens if I've modified the template prior to activating the plugin that has the auto-inserted block?**

The block will indeed _not_ be auto-inserted in that case; we’re considering showing some kind of notification in the UI though. This is covered in [this comment](https://github.com/WordPress/gutenberg/issues/39439#issuecomment-1150278043).
